### ISSUE TITLE
16 bit precision in Trainer leading to NaN

### ISSUE THREAD

**Issue by runyournode** [Lightning-AI/pytorch-lightning#20619]
### Description & Motivation

In Trainer, when using lightning_precision other than 32-true, small float values (~ > 1e-7 - 1e-8 ?) can be round up to zero (in input batches, model parameters and its gradients, optimizer hyper-parameters) leading to NaN. 

This does not raise any warning or error during training as the training loop ignore NaN. 
After training, in inference mode, all output can be NaN, (which makes it a poor model).

After debugging for several days, I found that I had issue in my parameter weights (both from blank initialization or from pre-trained) but also in the very commonly used Adam optimizer (with default eps=1e-8 it was updating all my parameters to NaN in the very first step, even when learning rate was set to 0).

### Pitch

I would suggest to add a caveat in the documentation to warn the user that one has to deal with every possible small values in its model, optimizer and input data.
Also some default warnings when autocasts to lower-precision lead to over/under -flow would be nice.
